File: 2024/day-11/main.ex
defmodule PlutonianPebbles do
  def solve(x, t, dp) do
    key = {x, t}

    if Map.has_key?(dp, key) do
      Map.get(dp, key)
    else
      result =
        cond do
          t == 0 ->
            1

          x == 0 ->
            solve(1, t - 1, dp)

          rem(String.length(Integer.to_string(x)), 2) == 0 ->
            dstr = Integer.to_string(x)
            mid = div(String.length(dstr), 2)
            left = String.slice(dstr, 0, mid) |> String.to_integer()
            right = String.slice(dstr, mid, String.length(dstr)) |> String.to_integer()
            solve(left, t - 1, dp) + solve(right, t - 1, dp)

          true ->
            solve(x * 2024, t - 1, dp)
        end

      dp = Map.put(dp, key, result)
      result
    end
  end

  def solve_all(d, t) do
    Enum.reduce(d, 0, fn x, sum ->
      dp = %{}
      sum + solve(x, t, dp)
    end)
  end

  def main() do
    case File.read("input.txt") do
      {:ok, data} ->
        d = data |> String.trim() |> String.split(~r/\s+/) |> Enum.map(&String.to_integer/1)

        dp = %{}
        result25 = solve_all(d, 25)
        result75 = solve_all(d, 75)

        IO.puts("Result after 25 blinks: #{result25}")
        IO.puts("Result after 75 blinks: #{result75}")

      {:error, reason} ->
        IO.puts("Error reading file: #{reason}")
    end
  end
end

PlutonianPebbles.main()

File: 2024/day-11/main.dart
import 'dart:io';
import 'dart:convert';

class Memoization {
  final Map<String, int> _cache = {};

  int solve(int x, int t) {
    final key = '$x,$t';
    if (_cache.containsKey(key)) {
      return _cache[key]!;
    }

    int result;
    if (t == 0) {
      result = 1;
    } else if (x == 0) {
      result = solve(1, t - 1);
    } else if (x.toString().length % 2 == 0) {
      final dstr = x.toString();
      final mid = dstr.length ~/ 2;
      final left = int.parse(dstr.substring(0, mid));
      final right = int.parse(dstr.substring(mid));
      result = solve(left, t - 1) + solve(right, t - 1);
    } else {
      result = solve(x * 2024, t - 1);
    }

    _cache[key] = result;
    return result;
  }
}

int solveAll(List<int> d, int t) {
  final memo = Memoization();
  return d.map((x) => memo.solve(x, t)).reduce((a, b) => a + b);
}

void main() {
  // Read input from input.txt
  final data = File('input.txt').readAsStringSync().trim();
  final d = data.split(RegExp(r'\s+')).map(int.parse).toList();

  // Calculate results for 25 and 75 blinks
  final result25 = solveAll(d, 25);
  final result75 = solveAll(d, 75);

  print('Result after 25 blinks: $result25');
  print('Result after 75 blinks: $result75');
}
